src/abstractdatatable/io.jl

```diff
@@ -195,73 +195,90 @@ end
+struct DataTableStream{T}
+    columns::T
+    header::Vector{String}
 end
+DataTableStream(dt::DataTable) = DataTableStream(Tuple(dt.columns), map(String, names(dt)))
 
-# DataTable as a Data.Source
+# DataTable Data.Source implementation
+function Data.schema(dt::DataTable)
+    return Data.Schema(Type[eltype(A) for A in dt.columns],
+                        map(string, names(dt)), length(dt) == 0 ? 0 : length(dt.columns[1]))
+end
+
+Data.isdone(source::DataTable, row, col, rows, cols) = row > rows || col > cols
 function Data.isdone(source::DataTable, row, col)
-    rows, cols = size(source)
-    return row > rows || col > cols
+    cols = length(source)
+    return Data.isdone(source, row, col, cols == 0 ? 0 : length(dt.columns[1]), cols)
 end
 
 Data.streamtype(::Type{DataTable}, ::Type{Data.Column}) = true
 Data.streamtype(::Type{DataTable}, ::Type{Data.Field}) = true
 
-Data.streamfrom{T}(source::DataTable, ::Type{Data.Column}, ::Type{T}, col) =
-    (A = source.columns[col]::AbstractVector{T}; return A)
-Data.streamfrom{T}(source::DataTable, ::Type{Data.Field}, ::Type{T}, row, col) =
-    (A = source.columns[col]::AbstractVector{T}; return A[row]::T)
+@inline Data.streamfrom(source::DataTable, ::Type{Data.Column}, ::Type{T}, row, col) where {T} = source[col]
+@inline Data.streamfrom(source::DataTable, ::Type{Data.Field}, ::Type{T}, row, col) where {T} = source[col][row]
+
+# DataTable Data.Sink implementation
+Data.streamtypes(::Type{DataTable}) = [Data.Column, Data.Field]
+Data.weakrefstrings(::Type{DataTable}) = true
 
-# DataTable as a Data.Sink
-allocate{T}(::Type{T}, rows) = Vector{T}(rows)
-allocate{T}(::Type{Vector{T}}, rows) = Vector{T}(rows)
+allocate(::Type{T}, rows, ref) where {T} = Vector{T}(rows)
+allocate(::Type{T}, rows, ref) where {T <: Union{WeakRefString, Null}} = WeakRefStringArray(ref, T, rows)
 
-function DataTable{T <: Data.StreamType}(sch::Data.Schema,
-                                         ::Type{T}=Data.Field,
-                                         append::Bool=false)
-    rows, cols = size(sch)
-    rows = max(0, T <: Data.Column ? 0 : rows) # don't pre-allocate for Column streaming
-    columns = Vector{Any}(cols)
+# Construct or modify a DataTable to be ready to stream data from a source with a schema of `sch`
+function DataTable(sch::Data.Schema{R}, ::Type{S}=Data.Field,
+                    append::Bool=false, args...; reference::Vector{UInt8}=UInt8[]) where {R, S <: Data.StreamType}
     types = Data.types(sch)
-    for i = 1:cols
-        columns[i] = allocate(types[i], rows)
+    if !isempty(args) && args[1] isa DataTable && types == Data.types(Data.schema(args[1]))
+        # passing in an existing DataTable Sink w/ same types as source (as indicated by `sch`)
+        sink = args[1]
+        sinkrows = size(Data.schema(sink), 1)
+        if append && (S == Data.Column || !R) # are we appending and either column-streaming or there are an unknown # of rows
```

Comment:
This line is still 130 chars... We can allow for some tolerance beyond 92 chars, but please, not that much! ;-)

Reply:
Woops, I commented on an outdated commit, but the point still applies to some comments even after the latest commit.

Reply:
this line specifically has been updated; there are other commented lines that are longer; are they ok? or should I break the comments up too?

Reply:
updated